### PR TITLE
feat: fetch new snapshots by entity timestamp

### PR DESCRIPTION
**Pull request by pedrotambo**
This PR allows processing snapshots with entities with `localTimestamp` (old ones) or `entityTimestamp`. It uses the new types in `common-schemas`.

PR dependency:  https://github.com/decentraland/common-schemas/pull/148

### CODE REVIEW

File: src/file-processor.ts [hunk 1]
```diff
@@ -41,13 +40,21 @@ export async function* processDeploymentsInFile(
...
   for await (const line of processLineByLine(stream)) {
     const theLine = line.trim()
     if (theLine.startsWith('{') && theLine.endsWith('}')) {
-      const deployment = coerceEntityDeployment(JSON.parse(theLine))
-      if (deployment) {
-        yield deployment
+      const parsedLine = JSON.parse(theLine)
+      if (SnapshotSyncDeployment.validate(parsedLine)) {
+        yield parsedLine
+      } else if (PointerChangesSyncDeployment.validate(parsedLine)) {
```
Comment (agusaldasoro) [inline src/file-processor.ts:50]: Are this the same entity type of the old snapshots?

File: src/index.ts [hunk 2]
```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
...
+
+        // 2.2. open the snapshot file and process line by line
+        const deploymentsInFile = processDeploymentsInFile(hash, components)
+        components.processedSnapshots.startStreamOf(hash)
+        let numberOfStreamedEntities = 0
+        for await (const deployment of deploymentsInFile) {
+
+          const deploymentTimestamp = 'entityTimestamp' in deployment ? deployment.entityTimestamp : deployment.localTimestamp
+
+          if (deploymentTimestamp >= genesisTimestamp) {
```
Comment (agusaldasoro) [inline src/index.ts:212]: do you remember why is this? could we add a comment here?

**Review by agusaldasoro — APPROVED** [review]

File: src/file-processor.ts (hunk 1 above)
Comment (pedrotambo) [reply src/file-processor.ts:50]: Yes. This line should be removed in the future when there are no old snapshots.

File: package.json [hunk 3]
```diff
@@ -17,9 +17,10 @@
   "dependencies": {
     "@dcl/catalyst-storage": "^1.0.0-20220604235145.commit-892b669",
     "@dcl/hashing": "^1.0.0",
-    "@dcl/schemas": "^5.4.2",
+    "@dcl/schemas": "tag-fast-bootstrapping-5",
```
Comment (menduz) [inline package.json:20]: use fixed version before merging

File: src/client.ts [hunk 4]
```diff
@@ -38,15 +63,25 @@ export async function* fetchJsonPaginated<T>(
...
-): AsyncIterable<DeploymentWithAuthChain> {
+  fromTimestamp: number,
+  logger: ILoggerComponent.ILogger
+): AsyncIterable<PointerChangesSyncDeployment> {
   const url = new URL(
     `${server}/pointer-changes?sortingOrder=ASC&sortingField=local_timestamp&from=${encodeURIComponent(fromTimestamp)}`
   ).toString()
-  return fetchJsonPaginated(components, url, ($) => $.deltas, 'dcl_catalysts_pointer_changes_response_time_seconds')
+  for await (const deployment of fetchJsonPaginated(components, url, ($) => $.deltas, 'dcl_catalysts_pointer_changes_response_time_seconds')) {
+    if (PointerChangesSyncDeployment.validate(deployment)) {
```
Comment (menduz) [inline src/client.ts:76]: Will this work for legacy pointer-changes?
```suggestion
    if (PointerChangesSyncDeployment.validate(deployment)) {
```

File: src/file-processor.ts [hunk 5]
```diff
@@ -41,13 +40,21 @@ export async function* processDeploymentsInFile(
...
+        yield parsedLine
+      } else {
+        const errors = {
+          ...SnapshotSyncDeployment.validate.errors,
+          ...PointerChangesSyncDeployment.validate.errors
+        }
+        console.error('ERROR: Invalid entity deployment in snapshot file', parsedLine, errors)
```
Comment (menduz) [inline src/file-processor.ts:57]: Use logger?
```suggestion
        console.error('ERROR: Invalid entity deployment in snapshot file', parsedLine, errors)
```

File: src/index.ts [hunk 6]
```diff
@@ -14,16 +15,57 @@ import {
...
 export { metricsDefinitions } from './metrics'
 export { IDeployerComponent } from './types'
+export { createProcessedSnapshotsComponent } from './processed-snapshots'
 
 if (parseInt(process.version.split('.')[0]) < 16) {
   const { name } = require('../package.json')
   throw new Error(`In order to work, the package ${name} needs to run in Node v16 or newer to handle streams properly.`)
 }
 
+async function downloadProfileAvatars(
```
Comment (menduz) [inline src/index.ts:30]: It is really hard to tell what this function does if we don't read to the name `downloadProfileAvatars` we may create an issue for it to tackle it later?

File: src/index.ts [hunk 7]
```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
...
+          options.tmpDownloadFolder,
+          [options.contentServer],
+          new Map(),
+          options.requestMaxRetries,
+          options.requestRetryWaitTime
+        )
+
+        // 2.2. open the snapshot file and process line by line
+        const deploymentsInFile = processDeploymentsInFile(hash, components)
+        components.processedSnapshots.startStreamOf(hash)
```
Comment (menduz) [inline src/index.ts:206]:
```suggestion
        await components.processedSnapshots.startStreamOf(hash)
```

File: src/index.ts (hunk 2 above)
Comment (menduz) [reply src/index.ts:212]: the genesisTimestamp is used in reconnections. So let's say you synchronize with a catalyst correctly up until NOW(1). then the remote catalyst gets restarted, that will trigger a re-synchronization from scratch locally. The genesisTimestamp may be set to NOW(1) to ignore all the history up until that moment.

The same logic should be applied for pointer-changes

File: src/index.ts [hunk 8]
```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
...
+          }
         }
-      }
-    } finally {
-      if (options.deleteSnapshotAfterUsage !== false) {
-        try {
-          await components.storage.delete([hash])
-        } catch (err: any) {
-          logs.error(err)
+        components.processedSnapshots.endStreamOf(hash, numberOfStreamedEntities)
```
Comment (menduz) [inline src/index.ts:225]:
```suggestion
        await components.processedSnapshots.endStreamOf(hash, numberOfStreamedEntities)
```

File: src/index.ts [hunk 9]
```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
...
       }
     }
+    greatestProcessedDeploymentTimestampFromSnapshots = lastIncludedDeploymentTimestamp > greatestProcessedDeploymentTimestampFromSnapshots
+      ? lastIncludedDeploymentTimestamp : greatestProcessedDeploymentTimestampFromSnapshots
   }
 
-  // 3. fetch the /pointer-changes of the remote server using the last timestamp from the previous step
+  logs.info('End streaming snapshots.', { greatestProcessedTimestamp: greatestProcessedDeploymentTimestampFromSnapshots })
+  // 3. fetch the /pointer-changes of the remote server using the last timestamp from the previous step with a grace period of 20 min
+  let localTimestampFromWhichFetchPointerChanges = greatestProcessedDeploymentTimestampFromSnapshots - ms('20m')
```
Comment (menduz) [inline src/index.ts:242]: Is it worth to add a dependency for a constant?
```suggestion
  let localTimestampFromWhichFetchPointerChanges = greatestProcessedDeploymentTimestampFromSnapshots - 20 * 60_000
```

File: src/index.ts [hunk 10]
```diff
@@ -132,79 +160,102 @@ export async function downloadEntityAndContentFiles(
...
-    const pointerChanges = fetchPointerChanges(components, options.contentServer, greatestProcessedTimestamp)
-    for await (const rawDeployment of pointerChanges) {
-      const deployment = coerceEntityDeployment(rawDeployment)
-      if (!deployment) continue
-      // selectively ignore deployments by localTimestamp
+    const pointerChanges = fetchPointerChanges(components, options.contentServer, localTimestampFromWhichFetchPointerChanges, logs)
+    for await (const deployment of pointerChanges) {
+
+      // selectively ignore deployments by entityTimestamp
       if (deployment.localTimestamp >= genesisTimestamp) {
```
Comment (menduz) [inline src/index.ts:249]: typo? is this covered in tests for compatibility?
```suggestion
      if (deployment.entityTimestamp >= genesisTimestamp) {
```

File: src/processed-snapshots.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      const snapshotWasAlreadyProcessed = processedSnapshotHashes.has(snapshotHash)
+      const replacedHashesWereAlreadyProcessed = replacedHashes.length > 0 && replacedHashes.every((h) => processedSnapshotHashes.has(h))
+
+      if (!snapshotWasAlreadyProcessed && replacedHashesWereAlreadyProcessed) {
+        await components.processedSnapshotStorage.saveProcessed(snapshotHash)
+      }
+
+      return !isBeingStreamed && !wasStreamed && !(snapshotWasAlreadyProcessed || replacedHashesWereAlreadyProcessed)
+    },
+    startStreamOf(snapshotHash: string) {
```
Comment (menduz) [inline src/processed-snapshots.ts:44]: I can think many ways in which having this being awaitable will be helpful in the future and won't hurt at this moment
```suggestion
    async startStreamOf(snapshotHash: string) {
```

File: src/types.ts [hunk 12]
```diff
@@ -117,4 +119,16 @@ export type CatalystDeploymentStreamOptions = DeployedEntityStreamOptions & {
...
+}
+
+export type IProcessedSnapshotStorageComponent = {
+  processedFrom(snapshotHashes: string[]): Promise<Set<string>>
+  saveProcessed(snapshotHash: string): Promise<void>
+}
+
+export type IProcessedSnapshotsComponent = {
+  shouldStream(snapshotHash: string, replacedSnapshotHashes?: string[]): Promise<boolean>
+  startStreamOf(snapshotHash: string): void
```
Comment (menduz) [inline src/types.ts:131]:
```suggestion
  startStreamOf(snapshotHash: string): Promise<void>
```

File: src/client.ts (hunk 4 above)
Comment (pedrotambo) [reply src/client.ts:76]: Hmmmm don't know. Anyways, it's just "moved" logic. That validation was taking place with the type `DeploymentWithAuthChain` which is equivalent to `PointerChangesSyncDeployment`.

File: src/index.ts (hunk 7 above)
Comment (pedrotambo) [reply src/index.ts:206]: That method is not a promise 🤔. Should we add it anyways?

File: src/index.ts (hunk 8 above)
Comment (pedrotambo) [reply src/index.ts:225]: This method IS a promise, added. 😝

File: src/index.ts (hunk 10 above)
Comment (pedrotambo) [reply src/index.ts:249]: Fixed typo in the comments

File: src/index.ts (hunk 7 above)
Comment (pedrotambo) [reply src/index.ts:206]: Okay, just read the next comments. Added as a promise

File: src/index.ts (hunk 9 above)
Comment (pedrotambo) [reply src/index.ts]: only god knows. Removed.

File: src/index.ts (hunk 6 above)
Comment (pedrotambo) [reply src/index.ts:30]: https://github.com/decentraland/snapshots-fetcher/issues

File: package.json (hunk 3 above)
Comment (pedrotambo) [reply package.json:20]: Updated the PR base branch from `main` to `feat/fetch-new-snapshots` so let's block the other one

**Pull request closed without merging**